### PR TITLE
Try and make background thread do something useful

### PR DESCRIPTION
**Pull request by anxiousmodernman**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #![allow(unused_imports)]
 #![allow(unused_must_use)]
 #![allow(dead_code)]
+#![feature(proc_macro)]
```
Comment (anxiousmodernman) [inline src/lib.rs:8]: need this in lib.rs as well as main.rs

File: src/main.rs [hunk 2]
```diff
@@ -5,8 +5,10 @@
 #![allow(unused_imports)]
```
Comment (anxiousmodernman) [inline src/main.rs:5]: we might put `main.rs` under a `bin` folder and name is server.rs or something

```
cargo run --bin server
```

Comment (jeffmay) [reply src/main.rs:5]: I think we'll probably want to split this file up a little better. I had a commit to do that, but it has a bunch of conflicts, so I'll take a crack at it again after merging.

File: src/main.rs [hunk 3]
```diff
@@ -93,10 +96,23 @@ fn index(db: State<Arc<DB>>) -> String {
...
 
 #[get("/random")]
-fn random(db: State<Arc<DB>>) -> String {
+fn random(db: State<Arc<DB>>) -> Markup {
     let mut iter = db.iterator(IteratorMode::Start);
 
-    String::from("Hello, world! ")
+
+    html! {
+        h1 "Jeff May?:"
```
Comment (jeffmay) [inline src/main.rs:104]: Oh hai!

File: src/project.rs [hunk 4]
```diff
@@ -33,10 +41,18 @@ pub struct FileInfo {
 
 impl FileInfo {
     pub fn from_path(path: &Path) -> Result<FileInfo, io::Error> {
+        println!("GETTING ONE FILE NOW {:?}", path);
         let file = File::open(path)?;
+        // more dumbass heuristics to avoid scanning binaries
+        println!("len {}", file.metadata().unwrap().len());
+        if file.metadata().unwrap().len() > 5000000 {
+            return Ok(FileInfo::blank())
+        }
```
Comment (jeffmay) [inline src/project.rs:50]: I wonder if we can try to scan the first chunk of the file and see if it is utf-8 or something and if not, then bail.

File: src/project.rs [hunk 5]
```diff
@@ -51,20 +67,55 @@ impl FileInfo {
...
+    ELF,
+    Text,
+    Markdown,
+}
+
+impl Render for FileInfo {
+    fn render(&self) -> Markup {
+        html! {
+
+        }
```
Comment (jeffmay) [inline src/project.rs:91]: Any plan for this?

Comment (anxiousmodernman) [reply src/project.rs:91]: just to get it to compile...

File: src/project.rs [hunk 6]
```diff
@@ -51,20 +67,55 @@ impl FileInfo {
...
-        let entry = entry.unwrap();
-        println!("name: {:?}, path: {:?}", entry.file_name(), entry.path());
+        println!("something happening");
+        let entry = entry.expect("could not get dir entry");
+        if entry.file_type().is_dir() {
+            println!("skipping analysis for dir");
+            continue;
+        }
+        println!("more stuff");
+        let k = make_key!(proj.to_str().expect("proj to string failed"), batch_id, entry.path().to_str().expect("path to string failed"));
```
Comment (jeffmay) [inline src/project.rs:113]: maybe chain this across multiple lines?

File: src/project.rs [hunk 7]
```diff
@@ -51,20 +67,55 @@ impl FileInfo {
...
 
-        // naive impl:
-        // "get" file repr from database
+        let fi = FileInfo::from_path(entry.path()).expect("could not make fileinfo");
+        println!("fileinfo made");
+        let encoded: Vec<u8> = serialize(&fi, Infinite).unwrap();
+        db.put(k.0.as_slice(), encoded.as_slice());
```
Comment (jeffmay) [inline src/project.rs:118]: What are the other values in the tuple of `k`?

File: src/project.rs (hunk 4 above)
Comment (anxiousmodernman) [reply src/project.rs:50]: This is dumb but it's a crazy hard problem.

For ELF executables, we can use an existing parser library. nrc wrote one https://github.com/nrc/xmas-elf

Comment (jeffmay) [reply src/project.rs:50]: Interesting...

File: src/project.rs (hunk 7 above)
Comment (anxiousmodernman) [reply src/project.rs:118]: it's a struct, but the idea is the same: fixed width with static index to the field

**Pull request merged**